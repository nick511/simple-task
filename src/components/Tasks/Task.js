import React, { Component } from 'react'
import PropTypes from 'prop-types'

import './task.css'

class Task extends Component {
  state = {
    showEdit: false,
    showEditName: false,
    nameValue: '',
  }

  /* react */
  componentDidMount = () => {
    // switch to change name mode after task added
    if (this.props.name === '') {
      this.handleShowChangeName()
    }
  }

  setTextInputRef = element => {
    this.textInput = element
  };

  /* handle actions */
  handleShowEdit = (e) => {
    this.setState({ showEdit: true })
  }

  handleChangeStatus = (e) => {
    const { id, isCompleted, actions } = this.props
    actions.updateStatus(id, !isCompleted)
    this.setState({ showEdit: false })
  }

  handleShowChangeName = (e) => {
    this.setState({
      showEditName: true,
      showEdit: false,
      nameValue: this.props.name,
    }, () => { // after updated
      if (this.textInput) {
        this.textInput.focus()
      }
    })
  }

  handleChangeName = (e) => {
    if (this.state.nameValue.trim() === '') {
      return
    }

    const { id, actions } = this.props
    actions.updateName(id, this.state.nameValue.trim())
    this.setState({ showEditName: false })
  }

  handleInputChange = (e) => {
    this.setState({ nameValue: e.target.value })
  }

  render () {
    const { name, isCompleted = false } = this.props
    const { showEdit, showEditName, nameValue } = this.state

    let status = 'Active'
    let statusClass = 'task__status'
    let editStatus = 'Complete'
    let editStatusClass = 'task__edit__complete'

    if (isCompleted) {
      status = 'Completed'
      statusClass += ' is-completed'
      editStatus = 'Undo'
      editStatusClass = 'task__edit__undo'
    }

    return (
      <div className='task'>

        <div className='task__main' onClick={this.handleShowEdit}>
          <div className='task__title'>{name}</div>
          <div className={statusClass}>{status}</div>
        </div>

        { showEdit &&
          <div className='task__overlay task__edit'>
            <div>
              <a className={editStatusClass} onClick={this.handleChangeStatus}>{editStatus}</a>
            </div>
            <div>
              <a className='task__edit__edit' onClick={this.handleShowChangeName}>Edit</a>
            </div>
          </div>
        }

        { showEditName &&
          <div className='task__overlay task__edit-name'>
            <div>
              <input type='text' placeholder='Enter task name...' ref={this.setTextInputRef} value={nameValue} onChange={this.handleInputChange} />
            </div>
            <div>
              <a className='task__edit-name__save' onClick={this.handleChangeName}>Save</a>
            </div>
          </div>
        }

      </div>
    )
  }
}

Task.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  isCompleted: PropTypes.bool,
  actions: PropTypes.shape({
    updateStatus: PropTypes.func.isRequired,
    updateName: PropTypes.func.isRequired,
  }),
}

export default Task
