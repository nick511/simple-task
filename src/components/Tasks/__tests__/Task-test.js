import React from 'react'
import { shallow } from 'enzyme'

import Task from '../Task'

describe('<Task />', () => {
  let wrapper = null
  let task = null

  beforeEach(() => {
    task = {
      id: 1,
      name: 'test',
      actions: {
        updateStatus: jest.fn(),
        updateName: jest.fn(),
      }
    }
    wrapper = shallow(<Task {...task} />)
  })

  test('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('renders task title', () => {
    expect(wrapper.find('.task__title').text()).toEqual(task.name)
  })

  test('renders task status', () => {
    expect(wrapper.find('.task__status').text()).toEqual('Active')
    wrapper.setProps({ isCompleted: true })
    expect(wrapper.find('.task__status').text()).toEqual('Completed')
  })

  test('renders task__edit after clicking task', () => {
    expect(wrapper.find('.task__status').text()).toEqual('Active')
    wrapper.find('.task__main').simulate('click')
    expect(wrapper.exists('.task__edit')).toEqual(true)

    wrapper.find('.task__edit a').at(0).simulate('click')
    expect(wrapper.exists('.task__edit')).toEqual(false)
    expect(task.actions.updateStatus).toHaveBeenCalledTimes(1)
  })
})
