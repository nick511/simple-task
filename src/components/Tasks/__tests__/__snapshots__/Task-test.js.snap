// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task /> renders without crashing 1`] = `
<div
  className="task"
>
  <div
    className="task__main"
    onClick={[Function]}
  >
    <div
      className="task__title"
    >
      test
    </div>
    <div
      className="task__status"
    >
      Active
    </div>
  </div>
</div>
`;
